File: 2025-11-November-GFG-POTD/November 24 - Second Best Minimum Spanning Tree/Solution.java
class Solution {
    // DSU for Kruskal
    class DSU {
        int[] p, r;
        DSU(int n){
            p = new int[n];
            r = new int[n];
            for(int i = 0; i < n; i++) p[i] = i;
        }
        int find(int x){
            return p[x] == x ? x : (p[x] = find(p[x]));
        }
        boolean union(int a, int b){
            a = find(a); b = find(b);
            if(a == b) return false;
            if(r[a] < r[b]) p[a] = b;
            else if(r[b] < r[a]) p[b] = a;
            else { p[b] = a; r[a]++; }
            return true;
        }
    }

    // Build MST, optionally skipping exactly one edge
    int buildMST(int V, int[][] edges, int skipU, int skipV, int skipW){
        DSU d = new DSU(V);
        int sum = 0, used = 0;

        for(int[] e : edges){
            if(e[0] == skipU && e[1] == skipV && e[2] == skipW) continue;
            if(d.union(e[0], e[1])){
                sum += e[2];
                used++;
            }
        }
        return (used == V - 1) ? sum : Integer.MAX_VALUE;
    }

    public int secondMST(int V, int[][] edges) {

        // Sort edges by weight
        Arrays.sort(edges, (a, b) -> a[2] - b[2]);

        // Build the true MST
        int mstWeight = buildMST(V, edges, -1, -1, -1);
        if(mstWeight == Integer.MAX_VALUE) return -1; // no MST

        int secondBest = Integer.MAX_VALUE;

        // Try removing each MST edge one by one
        for(int[] e : edges){
            int w = buildMST(V, edges, e[0], e[1], e[2]);
            if(w > mstWeight) secondBest = Math.min(secondBest, w);
        }

        return (secondBest == Integer.MAX_VALUE ? -1 : secondBest);
    }
}

File: 2025-11-November-GFG-POTD/November 24 - Second Best Minimum Spanning Tree/Solution.py
class Solution:

    # DSU for Kruskal
    class DSU:
        def __init__(self, n):
            self.p = list(range(n))
            self.r = [0] * n

        def find(self, x):
            if self.p[x] != x:
                self.p[x] = self.find(self.p[x])
            return self.p[x]

        def union(self, a, b):
            a, b = self.find(a), self.find(b)
            if a == b:
                return False
            if self.r[a] < self.r[b]:
                self.p[a] = b
            elif self.r[b] < self.r[a]:
                self.p[b] = a
            else:
                self.p[b] = a
                self.r[a] += 1
            return True

    # Build MST skipping exactly one edge
    def buildMST(self, V, edges, su, sv, sw):
        d = self.DSU(V)
        total = 0
        used = 0

        for u, v, w in edges:
            if u == su and v == sv and w == sw:
                continue
            if d.union(u, v):
                total += w
                used += 1

        return total if used == V - 1 else float('inf')

    def secondMST(self, V, edges):

        edges.sort(key=lambda x: x[2])

        mstWeight = self.buildMST(V, edges, -1, -1, -1)
        if mstWeight == float('inf'):
            return -1

        secondBest = float('inf')

        for u, v, w in edges:
            w2 = self.buildMST(V, edges, u, v, w)
            if w2 > mstWeight:
                secondBest = min(secondBest, w2)

        return -1 if secondBest == float('inf') else secondBest
